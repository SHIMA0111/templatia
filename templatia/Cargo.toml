[package]
name = "templatia"
description = "A template library writedown and load the struct"
version = { workspace = true }
authors = { workspace = true }
license = { workspace = true }
keywords = ["parser", "template"]
categories = ["parsing", "template-engine"]
edition = "2024"
rust-version = "1.85.0"
readme = { workspace = true }
repository = { workspace = true }
homepage =  { workspace = true }
documentation = { workspace = true }

[dependencies]
thiserror = "2"

templatia-derive = { version = "0.0.2", path = "../templatia-derive", optional = true }
chumsky = { version = "0.11", optional = true }

[features]
derive = ["templatia-derive", "chumsky"]