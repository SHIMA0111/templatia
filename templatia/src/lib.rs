//! Minimal template trait and error types for templatia.
//!
//! This crate defines the core Template trait and its error type used by the
//! templatia-derive procedural macro.
//!
//! # Examples
//! ```rust
//! use templatia::Template;
//!
//! // A consumer typically uses the derive macro from `templatia-derive`.
//! // The example below demonstrates the trait surface only.
//! struct Mock;
//!
//! impl Template for Mock {
//!     type Error = templatia::TemplateError;
//!     type Struct = Mock;
//!
//!     fn to_string(&self) -> String { String::from("mock") }
//!     fn from_string(_s: &str) -> Result<Self::Struct, Self::Error> { Ok(Mock) }
//! }
//! assert_eq!(Mock.to_string(), "mock");
//! ```

#[cfg(feature = "derive")]
pub mod derive {
    pub use templatia_derive::*;
}

/// A trait for converting between a struct and its string template form.
///
/// Types implementing this trait can serialize themselves into a template
/// string and be reconstructed from a string.
///
/// # Associated Types
/// - Error: Concrete error type used by the implementation.
/// - Struct: The concrete struct type to be produced by `from_string`.
pub trait Template {
    /// Concrete error type for template parsing or formatting failures.
    type Error;
    /// The struct type created by `from_string`.
    type Struct;

    /// Converts the value into its template string representation.
    ///
    /// # Returns
    /// A `String` that represents `self` according to the template rules.
    ///
    /// # Examples
    /// ```rust
    /// use templatia::Template;
    ///
    /// struct S;
    /// impl Template for S {
    ///     type Error = templatia::TemplateError;
    ///     type Struct = S;
    ///     fn to_string(&self) -> String { "x".into() }
    ///     fn from_string(_: &str) -> Result<Self::Struct, Self::Error> { Ok(S) }
    /// }
    /// assert_eq!(S.to_string(), "x");
    /// ```
    fn to_string(&self) -> String;

    /// Parses an instance from a string.
    ///
    /// # Parameters
    /// - s: Source string to parse.
    ///
    /// # Returns
    /// On success, returns a constructed instance of `Self::Struct`.
    ///
    /// # Errors
    /// Returns `Self::Error` if the string cannot be parsed.
    ///
    /// # Examples
    /// ```rust
    /// use templatia::Template;
    ///
    /// struct S;
    /// impl Template for S {
    ///     type Error = templatia::TemplateError;
    ///     type Struct = S;
    ///     fn to_string(&self) -> String { String::new() }
    ///     fn from_string(_: &str) -> Result<Self::Struct, Self::Error> { Ok(S) }
    /// }
    /// let _ = S::from_string("").unwrap();
    /// ```
    fn from_string(s: &str) -> Result<Self::Struct, Self::Error>;
}

/// Errors produced by templatia operations.
///
#[derive(Debug, thiserror::Error)]
pub enum TemplateError {
    /// The same placeholder occurred multiple times with different values.
    ///
    /// # Parameters
    /// - placeholder: The placeholder name.
    /// - first_value: The first observed value.
    /// - second_value: The conflicting later value.
    #[error("Inconsistent values for placeholder '{placeholder}': found '{first_value}', and after face '{second_value}'")]
    InconsistentValues {
        placeholder: String,
        first_value: String,
        second_value: String,
    },
    /// A generic parse error message aggregated from the parser.
    #[error("Parse error: {0}")]
    Parse(String),
}