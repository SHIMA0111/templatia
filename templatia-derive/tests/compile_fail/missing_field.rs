use templatia::Template;
use templatia_derive::Template as TemplateDerive;

#[derive(TemplateDerive)]
#[templatia(template = "host={host}\nport={port}\nuser={username}")]
struct DbCfg { host: String, port: u16 }

fn main() {}
