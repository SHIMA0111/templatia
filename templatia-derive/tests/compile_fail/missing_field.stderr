error: DbCfg has no field named "username"
 --> tests/compile_fail/missing_field.rs:6:8
  |
6 | struct DbCfg { host: String, port: u16 }
  |        ^^^^^

warning: unused import: `templatia::Template`
 --> tests/compile_fail/missing_field.rs:1:5
  |
1 | use templatia::Template;
  |     ^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
